src/Foundation/NSUrlSessionHandler.cs

```diff
@@ -215,13 +215,18 @@ InflightData GetInflightData (NSUrlSessionTask task)
 					if (sessionHandler.inflightRequests.TryGetValue (task, out inflight))
 						return inflight;
 
+				// if we did not manage to get the inflight data, we either got an error or have been canceled, lets cancel the task, that will execute DidCompleteWithError
+				task.Cancel ();
 				return null;
 			}
 
 			public override void DidReceiveResponse (NSUrlSession session, NSUrlSessionDataTask dataTask, NSUrlResponse response, Action<NSUrlSessionResponseDisposition> completionHandler)
 			{
 				var inflight = GetInflightData (dataTask);
 
+				if (inflight == null)
+					return;
+
 				try {
 					// ensure that we did not cancel the request, if we did, do cancel the task
 					if (inflight.CancellationToken.IsCancellationRequested) {
```

Comment:
Well, I moved the task.Cancel there, to GetInflightData, at that point, the DidCompleteWithError handler will be executed, everything will be cleaned. AS per docs, the DidReceiveResponse gets called just when we got the headers (or at least that is what I understand from https://developer.apple.com/reference/foundation/nsurlsessiondatadelegate/1410027-urlsession?language=objc) .

What can happen is that we got the headers AND suddenly (if even possible) then got canceled, so we check there. Next time we will deal with the data from DidReceiveData (https://developer.apple.com/reference/foundation/nsurlsessiondatadelegate/1411528-urlsession?language=objc) in that case, we will get data BUT we also have to check if the inflight is null to ensure that we did not get canceled in the middle of the data being received, if we did get canceled, again, DidCompleteWithError will be taking care of everything, else, we wait for another DidReceiveData, and again, we could have been canceled again.

So, the check for the inflight data to be null has to be done in each of the Handlers that are receiving information from the task, check it, ensure that we got the data, if not, return.

I moved the task cancelation to the GetInflightData, since it makes no sense to do that check, cancel every time, but I don't see a "nice" way to avoid the check in the handlers since we could be canceled in the middle of each of the events.

So, what I'm thinking is that in a multithreaded situation we could have:

Handler is called -> Get Canceled -> Continue with handler

So ideally we should do

Handler is called -> Get Canceled -> Check inflight data -> Continue with handler

and getting the inflight data already will cancel the NSUrlsTask if the inflight data was cleaned. That data can be cleaned by several reasons:

1. Got an error form the os, and the handler is cleaning stuff.
2. CancellationToken was used (line 177)
3. The content was disposed during the DidReciveResponse handler (line 239)
4. An exception occurred in the DidReceiveResponse (line 277)